### PR TITLE
Completed the Third Phase Targets for GSOC

### PR DESCRIPTION
**Pull request by im-prakher**
Targets achieved in this PR -
- Visualized the polygons as they should have been, also modified the tests for that in [#0dcd13f](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/0dcd13fb3bf1bdb1baecf0f03f6d60829eefd906). 

- Implemented Data Validation and Improved Error Logging & Handling in [#629583f](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/629853feced8f9704d8449ea7cb46fc9d7051c31) and [#1c39008](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/1c3900808502a481ae6f9a8b489c9092817b808a)

- Improved SheetToCSV functionality in [#ca9adac](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/ca9adacfd4a3c16bfc7db28dd7e7a570cf2f1f32)

- API Key Validation done in [#6842d16](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/6842d16a622a4f75bbbe3df05665fd400be9a495)

- Changed the Docker container ports for visualization [#ed56e5c](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/ed56e5c9234d6d22324b19981fef462fc7540e11)

- Completed the documentation of YABA App [#8b3c15c](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/8b3c15ce1a811922d19c7d723ce3a04641694c81)

Also before merging this PR please merge #63.
Kindly review @dlebauer @KristinaRiemer  @Chris-Schnaufer

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,25 +53,187 @@ docker-compose run --rm bety initialize
...
-#### Once initialization and synchronization of bety database is done,the below command can be directly used to start the app.(Skip the "Setup BETYdb" step) 
 ```sh
 # Bring up full stack
 docker-compose up
 ```
-## How to hit the YABA API endpoints
+
+## Steps to Upload Metadata
+
+**1.** Enter your API key
```
Comment (KristinaRiemer) [inline README.md]: If just the numbers with the following period are here, instead of bolding them, it will make it into an actual list. [Markdown list syntax](https://www.markdownguide.org/basic-syntax/#ordered-lists). I showed how to do this for the first list item below: 

```suggestion
1. Enter your API key
```

**Comment by KristinaRiemer** [comment]
@im-prakher whether I put in my real API key from [the TERRA REF Bety](https://terraref.ncsa.illinois.edu/bety/) or some random characters, I get this error. 
Screen Shot 2020-08-31 at 8 03 26 AM

And this is the pop up window after I put a key in and hit the start button. What is supposed to go in as `[object Object]`? 
Screen Shot 2020-08-31 at 8 03 13 AM

**Comment by im-prakher** [comment]
@KristinaRiemer  Bety database image currently doesn't contain apikey for any user so it couldn't find it for you. Meanwhile, I will make a commit so that it displays an appropriate message for this. 

Here's the issue for this
[#46](https://github.com/PecanProject/BETYdb-YABA/issues/46)

**Comment by im-prakher** [comment]
@KristinaRiemer  Also, you can insert the API key for any user if you want by using these comands,
1. `docker exec  -ti <container_id_for_postgres> bin/bash`
2. `su postgres`
3. `psql`
To get potgres container id you can use `docker ps`,
You will now be accessing the postgres database 
4. Switch user `\c bety`
5. Run the query to set the apikey for a user in `users` table
`update users set apikey = 'oxDrvOozw+TVV0GKYOcTPooBuHIQymBaVxV4BbRo' where name = 'user 1';`
You can set the `apikey= oxDrvOozw+TVV0GKYOcTPooBuHIQymBaVxV4BbRo` for `user1`
as it the apikey column has some constraints of its own

File: app/yaba.yaml [hunk 2]
```diff
@@ -129,9 +140,10 @@ paths:
                     sitename,city,state,country,notes,greenhouse,geometry,time_zone,soil,soilnotes
                     'MAC Field Scanner Season 8 Range 1 Column 1 W','Maricopa','Arizona','USA','This site represents the the whole wo...',...,'America/Phoenix','soil','soilnotes,
                   
-                  externalValue: 'https://docs.google.com/spreadsheets/d/1BPkZPfynZI9DB9bRJfa335uC_uNjST5jHVf-XUgSVvM/edit#gid=347205052'
-      
+                  externalValue: 'https://docs.google.com/spreadsheets/d/1BPkZPfynZI9DB9bRJfa335uC_uNjST5jHVf-XUgSVvM/edit#gid=347205052'
```
Comment (Chris-Schnaufer) [inline app/yaba.yaml]: This link, and others like it, should not be in edit mode: `... /edit=#gid=...`

File: interface/src/components/homepage.js [hunk 3]
```diff
@@ -1,8 +1,35 @@
...
+    async updateUser(){
+        let apikey= document.getElementById("api-key").value;
+        if(apikey !== ""){
+            let user= await getUserData(apikey);
+            if(user.hasOwnProperty("error")){
+                return alert('Encountered error: '+ user.error);
+            }
+            if(Object.keys(user).length === 0)
+                return alert('User not found. Please enter a valid API key');
+            let isValid = window.confirm(`Please confirm, are you ${user}?`);
```
Comment (Chris-Schnaufer) [inline interface/src/components/homepage.js:21]: I don't know the design behind this so my next comment may not be meaningful: Is this confirmation needed? It seems that instead of asking for confirmation that it might be better for the user's name to be displayed prominently on subsequent pages.

**Comment by im-prakher** [comment]
@KristinaRiemer This error is now solved in [#41db3ae](https://github.com/PecanProject/BETYdb-YABA/pull/69/commits/41db3aedb89678f116f79f560d519f281952ed56)

File: visualization/pg_joins.js [hunk 4]
```diff
@@ -52,40 +52,67 @@ const fetch_specieId= async (client, specie)=>{
...
+}
 /*We just need the data after performing joins on some tables, for only visualization purposes not insertion,
  therefore a ROLLBACK is performed at end of this function*/
 
-const getCultivarSites=async (cultivars, sites_cultivars, sites) => {
+const getCultivarSites=async (cultivars, sites_cultivars, sites, geoJSON) => {
   const client = await pool.connect()
   try{
     await client.query('BEGIN')
+    let t1= ((await client.query("SELECT (UTC_NOW() - interval '1 second') as utc_now")).rows[0]["utc_now"]);
```
Comment (Chris-Schnaufer) [inline visualization/pg_joins.js:76]: Is this sufficient amount of time into the past? Perhaps getting the system time before making the queries, caching and using it for these queries (here and elsewhere that use the same time specification) would more certain to capture all the data?

**Review by Chris-Schnaufer — CHANGES_REQUESTED** [review]

File: interface/src/components/homepage.js (hunk 3 above)
Comment (KristinaRiemer) [reply interface/src/components/homepage.js:21]: We do want this confirmation step at that point in time. Basically the user enters their Bety api key, and that can confirm that they put in the correct one because their user name in Bety should show up.

Comment (Chris-Schnaufer) [reply interface/src/components/homepage.js:21]: OK

File: README.md (hunk 1 above)
Comment (im-prakher) [reply README.md]: Changed in future commit

File: app/yaba.yaml (hunk 2 above)
Comment (im-prakher) [reply app/yaba.yaml]: Converted edit to view mode

File: visualization/pg_joins.js (hunk 4 above)
Comment (im-prakher) [reply visualization/pg_joins.js:76]: Yeah! I checked that and it is sufficient enough. Also, getting the system time before making queries is sometimes the same as the time taken while performing them which is not making any difference.

**Comment by im-prakher** [comment]
@dlebauer You can merge this PR and #63 as well

**Comment by KristinaRiemer** [comment]
@im-prakher I just went through this PR to see how everything worked. I'm documenting all the steps I took here, in case I did something wrong and also because there were a few issues. 

I removed all of the existing containers, including the database, to prevent duplicate data errors with `docker system prune -a`. Then, while on the `containerize` branch I followed all [the README instructions](https://github.com/PecanProject/BETYdb-YABA/blob/8b3c15ce1a811922d19c7d723ce3a04641694c81/README.md) to get the interface and visualization containers running. I then switched over to this `validate` branch and ran `docker-compose up` to get these containers. 

I opened the app and put in the API key after following [these instructions](https://github.com/PecanProject/BETYdb-YABA/pull/69#issuecomment-683869813), and that went great. 

The three visualization pages also looked good except for one thing. When I clicked on the plots in the experiment viz, they said they were locations at range 39, but the input data in experiments_sites.csv said range 20. (See screenshot below) Why are those different? 
Screen Shot 2020-09-24 at 8 11 34 AM

Then after the visualization pages, I got this not iterable error on the following page: 
Screen Shot 2020-09-24 at 8 13 00 AM

Lastly, I checked the data validation. I redid all the steps from above to remove and recreate the containers to make sure there was no example data in the database already. I then removed the first column (`name`) from the cultivars.csv to see if the validation would catch that. I uploaded all the data files, but the validate page said all tables were successfully validated, though I expected an error about the cultivars table. Should there have been an error? 

There was an error on the next page for the cultivars visualization: 
Screen Shot 2020-09-24 at 9 04 24 AM

**Comment by im-prakher** [comment]
@KristinaRiemer The difference between the ranges of plots in viz part is due to the different sitename in the experiments table and shapefiles, e.g.
In the experiments table, `MAC Field Season 8 Range 20 Column 1 W`
But in shapefile, it is `Range 39 Column 1` 
This is happening as the table contains East and West for 2 polygons with the same range and column but in shapefile, there is no east and west, only ranges and columns so 2 polygons with the same range and column in experiments will have different range and column in shapefile.
And the viz part is using the shapefile for displaying the plots name.

We have discussed this type of situation during the GSoC meeting and we planned to get the sitename for plots from shapefile

**Comment by im-prakher** [comment]
The iterable error that occurred for three tables is because the insert_sites route isn't working properly and all these tables are dependent on the sites table which must be uploaded before uploading them. Once the sites route is fixed, it will work.

**Comment by im-prakher** [comment]
I'll look into the validation error, it must be something about the yaba_api code in python

File: app/Meta.py [hunk 5]
```diff
@@ -34,7 +34,7 @@ def save_tempFile(File):
     File.save(UPLOAD_PATH)
     return None
 
-def insert_experiments(username,fileName):
+def insert_experiments(username,fileName,status=True):
```
Comment (Chris-Schnaufer) [inline app/Meta.py:35]: Please document all the parameters that don't have any documentation yet

File: app/Meta.py [hunk 6]
```diff
@@ -83,12 +91,13 @@ def insert_experiments(username,fileName):
...
-    except Exception :
+    except Exception as e:
         # Logs the error appropriately
         logging.error(traceback.format_exc())
-        return 410     
+        msg = {'Message' : 'Error' + str(e)}
+        return make_response(jsonify(msg), 410)
      
-def insert_sites(fileName,shp_file,dbf_file,prj_file,shx_file):
+def insert_sites(fileName,shp_file,dbf_file,prj_file,shx_file,status=True):
```
Comment (Chris-Schnaufer) [inline app/Meta.py:98]: Please ensure all the parameters are documented here and elsewhere as well

File: app/yaba.yaml [hunk 7]
```diff
@@ -22,6 +22,11 @@ paths:
           schema:
             type: string
           description: Username uploading the csv
+        - in: query
+          name: status
+          schema:
+            type: boolean
+          description: Status for uploading.
```
Comment (Chris-Schnaufer) [inline app/yaba.yaml:29]: From this description I'm not sure how to use this parameter. Are you able to reword it to include expected effects? [Perhaps it doesn't perform the uploading, just checks that it can be done? I don't know]

File: app/yaba.yaml [hunk 8]
```diff
@@ -101,6 +106,12 @@ paths:
       operationId: Meta.insert_sites
       tags:
         - sites
+      parameters:
+        - in: query
+          name: status
+          schema:
+            type: boolean
+          description: Status for uploading.
```
Comment (Chris-Schnaufer) [inline app/yaba.yaml:114]: Same comment here as for above `status` description. The description for `status` can also be updated elsewhere

File: interface/src/components/gmap.js [hunk 9]
```diff
@@ -1,62 +1,166 @@
...
+import { bbox, geojsonType } from '@turf/turf'
 
 
 class GMap extends Component{    
     constructor(props) {
         super(props);
         this.state={
-            position: [51.505, -0.09]
-      }    
+            position: [ 33.056702, -112.046656 ],
```
Comment (Chris-Schnaufer) [inline interface/src/components/gmap.js:15]: Can `position` be a prop? It could also be stored in a cookie, HTML5 indexedDB (https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API), or other storage, instead of being hard coded to this spot.

File: interface/src/components/homepage.js [hunk 10]
```diff
@@ -1,8 +1,35 @@
...
+            if(isValid === true){
+                this.props.setUser(user);
+                return this.props.history.push("/upload1");
+            }
+            return;
+        }
+        if(this.props.user !== undefined)
+        return this.props.history.push("/upload1");
+        return alert('Please enter a valid API key');
```
Comment (Chris-Schnaufer) [inline interface/src/components/homepage.js]:
```suggestion
        if(this.props.user !== undefined) {
            return this.props.history.push("/upload1");
        }
        return alert('Please enter a valid API key');
```

Comment (Chris-Schnaufer) [inline interface/src/components/homepage.js]: IMO it would be better to have an Error Message component display the error, rather than an alert. This way focus can also be set to the control that needs better input and/or other UI decorations could also be set to highlight whats wrong

File: interface/src/components/navbar.js [hunk 11]
```diff
@@ -1,19 +1,16 @@
...
-        }
-        console.log(x);
-    }
-   
+const navbar= ({ user })=>{
+  const User=()=>{
+      if(user !== undefined){
+          return (
+            <div>
+                <img src="/user.svg" className="user-img" alt="User"/>
```
Comment (Chris-Schnaufer) [inline interface/src/components/navbar.js:9]: Nice use of SVG file!

File: docker-compose.yml [hunk 12]
```diff
@@ -48,11 +48,38 @@ services:
       - 6001:6000
     depends_on:
       - yaba_api
+  
+  yaba_visualization:
```
Comment (Chris-Schnaufer) [inline docker-compose.yml]: Would it make sense to set environment variables here that reference the other YABA server addresses that can be accessed through ReactJS's `process.env`? This way only one place needs to change if the server information changes

File: interface/src/components/requests.js [hunk 13]
```diff
@@ -25,49 +27,62 @@ export function getVisualData(files, type, username){
...
+}
 
 export function getGeoJSON(file){
     const formData = new FormData();
+    formData.append('shp_file', file, file.name);
 
-    formData.append("shp_file", file);
-
-    let res= fetch("http://localhost:3001/getGeoJSON", {
+    let res= fetch("http://localhost:8008/getGeoJSON", {
```
Comment (Chris-Schnaufer) [inline interface/src/components/requests.js:68]: See comment in docker-compose file above for reference: this is an example of where having the host URI defined as an environment variable would be handy

File: interface/src/components/sheet-parser.js [hunk 14]
```diff
@@ -1,22 +1,21 @@
...
 async function parseToCsv(googleSheetLink){
     return await new Promise((resolve,reject)=>{
-        const Tabletop= require('tabletop')
-        let sheetID = parser(googleSheetLink)["key"];
-         Tabletop.init({
-            key: sheetID,
-            callback:(data) => { resolve(showInfo(data)) },
-            simpleSheet: true
-        })    
+        googleSheetLink= 'https://cors-anywhere.herokuapp.com/' + googleSheetLink;
```
Comment (Chris-Schnaufer) [inline interface/src/components/sheet-parser.js:5]: Would it make sense to have these sheets locally available instead of Google? Perhaps fetching them as part of the docker build process?

File: interface/src/components/upload1.js [hunk 15]
```diff
@@ -43,17 +44,21 @@ class Upload extends Component{
...
             this.props.getFiles(this.state)
             return this.props.history.push("/upload2");
         }
-        str.join(", ");
+        if(err_link.length!==0){
+             err_link=err_link.join(", ")
+             alert(`Please check your google sheet link for ${err_link} files`);
+        }
+        str= str.join(", ");
         alert(`You have not uploaded your ${str} files`);
```
Comment (Chris-Schnaufer) [inline interface/src/components/upload1.js:62]:
```suggestion
        alert(`Please uploaded your ${str} files before continuing`);
```
This is another spot where an error component may be handy instead of an alert

File: interface/src/components/upload2.js [hunk 16]
```diff
@@ -47,16 +48,21 @@ class Upload extends Component{
...
-           return this.props.history.push("/visual");
+           return this.props.history.push("/validate");
        }
-       str.join(", ");
+       if(err_link.length!==0){
+            err_link=err_link.join(", ")
+            alert(`Please check your google sheet link for ${err_link} files`);
+       }
+       str= str.join(", ");
        alert(`You have not uploaded your ${str} files`);
```
Comment (Chris-Schnaufer) [inline interface/src/components/upload2.js:66]: Same comments as in `upload1.js`

File: interface/src/components/validation.js [hunk 17]
```diff
@@ -1,13 +1,108 @@
...
+    return <Redirect
+            to= {{
+                pathname: this.state.path,
+                state: { message: this.state.message }
+            }}/>;
     }
+  return (
+    <div id="validation">
+      <div className='table'>
+        <h1 id='title'>Check weather your files are ready or not?</h1>
```
Comment (Chris-Schnaufer) [inline interface/src/components/validation.js]:
```suggestion
        <h1 id='title'>Check whether your files are ready or not?</h1>
```

File: interface/src/components/validation.js [hunk 18]
```diff
@@ -1,13 +1,108 @@
...
+import { uploadFiles } from './requests'
+import './stylesheets/validation.css'
 
 class Validation extends Component{
-    constructor(props){
-      super(props);
+  constructor(props) {
+    super(props) //since we are extending class Table so we have to use super in order to override Component class constructor
+    this.state = { //state is by default an object
+       onLoad: false,
```
Comment (Chris-Schnaufer) [inline interface/src/components/validation.js:10]: Perhaps rename this state variable to `hasLoaded` - I think the "onLoad" name is too close to an event's name

File: visualization/index.js [hunk 19]
```diff
@@ -1,9 +1,9 @@
 const express = require('express')
 const app = express()
-const port = 3001
-const { getCultivarSites, getExperimentSites, getTreatmentSites } = require('./pg_joins')
+const port = 8075
+const host= '0.0.0.0'
```
Comment (Chris-Schnaufer) [inline visualization/index.js]: See docker compose comment above: potentially a good place for environment variables

File: visualization/index.js [hunk 20]
```diff
@@ -13,7 +13,7 @@ const path= require('path')
 app.use(express.static('dist'))
 app.use(express.json())
 app.use(function (req, res, next) {
-  res.setHeader('Access-Control-Allow-Origin', 'http://localhost:3000');
+  res.setHeader('Access-Control-Allow-Origin', 'http://localhost:3001');
```
Comment (Chris-Schnaufer) [inline visualization/index.js]: See docker compose comment above: potentially a good place for an environment variable

**Review by Chris-Schnaufer — COMMENTED** [review]
Added some comments for improving documentation, and optional suggestions for code improvements

**Comment by KristinaRiemer** [comment]
@im-prakher great, thank you for the explanations about the plot names and iterable error! 

With respect to the validation error, the error shown in the screenshot is correct, I think. The problem is that the validation page said that all of the tables were valid, even though I had purposely changed the cultivars table so that it wasn't valid. The validation page should return that there is something wrong with the cultivars table, I believe?

File: app/Meta.py (hunk 5 above)
Comment (im-prakher) [reply app/Meta.py:35]: Will make this more clear in the description!

File: app/Meta.py (hunk 6 above)
Comment (im-prakher) [reply app/Meta.py:98]: Will make this more clear in the description!

File: app/yaba.yaml (hunk 7 above)
Comment (im-prakher) [reply app/yaba.yaml:29]: Will make this more clear in the description!

File: app/yaba.yaml (hunk 8 above)
Comment (im-prakher) [reply app/yaba.yaml:114]: Will make this more clear in the description!

File: interface/src/components/gmap.js (hunk 9 above)
Comment (im-prakher) [reply interface/src/components/gmap.js:15]: Yeah, we can try this. Actually, it's the default position [ 33.056702, -112.046656 ] for the map pointing to Maricopa, AZ.  Within a few seconds, the map displays the shapefile data so do we really need to make the user location as default?

File: interface/src/components/homepage.js (hunk 10 above)
Comment (im-prakher) [reply interface/src/components/homepage.js]: Actually, there is a component for errors ([interface/src/components/error.js](https://github.com/PecanProject/BETYdb-YABA/blob/688fcfe3f7e08535013586c06d0689ccb0f4d114/interface/src/components/error.js)) but we just want to overlay a message for an invalid API key rather redirecting to an error page

File: interface/src/components/navbar.js (hunk 11 above)
Comment (im-prakher) [reply interface/src/components/navbar.js:9]: Thank you

File: interface/src/components/sheet-parser.js (hunk 14 above)
Comment (im-prakher) [reply interface/src/components/sheet-parser.js:5]: This might make it difficult for users to get data through sheets as they have to rebuild the whole project each time if they have new or different sheets.

File: interface/src/components/upload1.js (hunk 15 above)
Comment (im-prakher) [reply interface/src/components/upload1.js:62]: Again, there is a component for errors ([interface/src/components/error.js](https://github.com/PecanProject/BETYdb-YABA/blob/688fcfe3f7e08535013586c06d0689ccb0f4d114/interface/src/components/error.js)) but we just want to overlay a message for not uploading some files rather redirecting to an error page

File: interface/src/components/upload2.js (hunk 16 above)
Comment (im-prakher) [reply interface/src/components/upload2.js:66]: Again, there is a component for errors ([interface/src/components/error.js](https://github.com/PecanProject/BETYdb-YABA/blob/688fcfe3f7e08535013586c06d0689ccb0f4d114/interface/src/components/error.js)) but we just want to overlay a message for an invalid API key rather redirecting to an error page

File: interface/src/components/validation.js (hunk 18 above)
Comment (im-prakher) [reply interface/src/components/validation.js:10]: Yeah, that sounds nice.

File: docker-compose.yml (hunk 12 above)
Comment (im-prakher) [reply docker-compose.yml]: Yeah, that's a great idea. Here a new issue for that #72 .

File: visualization/index.js (hunk 19 above)
Comment (im-prakher) [reply visualization/index.js]: Good idea, new issue #72

File: visualization/index.js (hunk 20 above)
Comment (im-prakher) [reply visualization/index.js]: Good idea, new issue #72

File: interface/src/components/requests.js (hunk 13 above)
Comment (im-prakher) [reply interface/src/components/requests.js:68]: New issue #72 for that

File: README.md [hunk 21]
```diff
@@ -108,7 +304,7 @@ curl -F "fileName=@input_files/citations_sites.csv"   \
      http://localhost:5001/yaba/v1/citations_sites
 ```
 
-## How to hit the Client Endpoints
+### Hit the Client Endpoints
```
Comment (Chris-Schnaufer) [inline README.md]:
```suggestion
### Accessing the Client Endpoints
```

**Review by Chris-Schnaufer — APPROVED** [review]
